File: 백준/Silver/1920. 수 찾기/수 찾기.py
'''
수 찾기.
n = int(input())
lst = list(map(int,input().split()))
m = int(input())
target = list(map(int,input().split()))

ans = []
for i in target:
    if i in lst:
        ans.append(1)
    else:
        ans.append(0)

for i in ans:
    print(i)

위 코드 시간초과 됨. 시간 복잡도 : O(N)

모든 정수의 범위 -2^31 <= A < 2^31

자연수 n,m 은 10만개이하.

이분탐색. 으로 시간을 줄임. 시간 복잡도 :  O(logN)
def binary_search(array, target, start, end):
    if start > end:
        return None
    mid = (start + end) // 2

    if array[mid] == target:
        return mid
    elif array[mid] > target:
        return binary_search(array, target, start, mid-1)
    else:
        return binary_search(array, target, mid+1, end)
'''
import sys
input = sys.stdin.readline

n = int(input())
lst = sorted(map(int, input().split()))

m = int(input())
target = map(int, input().split())


def bi(array, ta, start, end):
    if start > end:
        return 0
    mid = (start + end) // 2

    if array[mid] == ta:
        return 1
    elif array[mid] < ta:
        return bi(array, ta, mid + 1, end)
    else:
        return bi(array, ta, start, mid - 1)


for i in target:
    start = 0
    end = n - 1
    print(bi(lst, i, start, end))

File: 백준/Silver/1920. 수 찾기/수 찾기.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(bf.readLine());
        int[] A = new int[N];
        StringTokenizer st = new StringTokenizer(bf.readLine());
        for(int i=0; i<N; i++){
            A[i] = Integer.parseInt(st.nextToken());
        }
        //A 배열 정렬.
        Arrays.sort(A);
        // 입력값 하나씩 이분탐색
        int M = Integer.parseInt(bf.readLine());
        st = new StringTokenizer(bf.readLine());
        for(int i=0; i<M; i++){
            boolean find = false;
            int target = Integer.parseInt(st.nextToken());
            int start = 0;
            int end = A.length -1;
            while(start <= end){
                int mid = (start+end)/2;
                int mid_value = A[mid];
                if(mid_value > target){
                    end = mid - 1;
                } else if(mid_value < target){
                    start = mid+1;
                } else {
                    find = true;
                    break;
                }
            }
            if(find) System.out.println(1);
            else System.out.println(0);
        }

    }

}
